### PR TITLE
Added view validation and solution n+1 problems

### PR DESCRIPTION
**Pull request by cmolinan**
### ACTIVITIES DONE

- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user. 
#### Integration specs
- Use Capybara to write integration tests for each view in the project. 
  - User index page
  - user show page
  - User post index page
  - Post show page

### CODE REVIEW

File: spec/requests/user_spec.rb [hunk 1]
```diff
@@ -17,5 +17,10 @@
       @user.post_counter = -1
       expect(@user).to_not be_valid
     end
+
+    it 'PostCounter must be greater than or equal to zero' do
+      @user.post_counter = 7
+      expect(@user).to be_valid
+    end
   end
 end
```
Comment (AnselemOdims) [inline spec/requests/user_spec.rb]: - [ ] Good job adding some tests but please kindly include integration tests for `User` index and show page with the following requirements.

### Index Page
- Tests that assert you can see the username of all other users.
- Tests that assert you can see the profile picture for each user.
- Tests that assert you can see the number of posts each user has written.
- Tests that redirect to the user's show page when you click on a user.

### Show Page
* Tests that assert you can see the user's profile picture.
* Tests that assert you can see the user's username.
* Tests that assert you can see the number of posts the user has written.
* Tests that assert you can see the user's bio.
* Tests that assert you can see the user's first 3 posts.
* Tests that assert you can see a button that lets me view all of a user's posts.
* Tests that redirect to the post's show page when you click on a user's post.
* Tests that redirect to the user's post's index page when you click on a **see all posts**.

File: spec/requests/posts_specs.rb [hunk 2]
```diff
@@ -1,32 +1,42 @@
...
 
-    it 'render the correct page' do
-      get('/users/403/posts/1')
-      expect(response.body).to render_template('show')
+    it 'if comments counter is integer' do
+      @post.comments_counter = 8
+      expect(@post).to be_valid
     end
   end
 end
```
Comment (AnselemOdims) [inline spec/requests/posts_specs.rb]: - [ ] Good job adding some tests but please kindly include integration tests for `Post` index and show page with the following requirements.

### Index Page
* Tests that asserts you can see the user's profile picture.
* Tests that asserts you can see the user's username.
* Tests that asserts you can see the number of posts the user has written.
* Tests that asserts you can see a post's title.
* Tests that asserts you can see some of the post's body.
* Tests that asserts you can see the first comments on a post.
* Tests that asserts you can see how many comments a post has.
* Tests that asserts you can see how many likes a post has.
* Tests that asserts you can see a section for pagination if there are more posts than fit on the view.
* Tests that redirects to the post's show page when you click on a post.

### Show Page
* Tests that asserts you can see the post's title.
* Tests that asserts you can see who wrote the post.
* Tests that asserts you can see how many comments it has.
* Tests that asserts you can see how many likes it has.
* Tests that asserts you can see the post body.
* Tests that asserts you can see the username of each commentor.
* Tests that asserts you can see the comment each commentor left.

**Review by AnselemOdims — CHANGES_REQUESTED** [review]
### Hi @rebel216, 
You have been doing a tremendous job so far :ok_hand:
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
PR has a title and summary. :heavy_check_mark:
No linter errors :heavy_check_mark:

## Required Changes ♻️
- [ ] Please also check inline comments under the review

## Optional suggestions
_Comments with the [OPTIONAL] prefix are not a must to implement however, It's strongly recommended to work on them as they can make your code better._

## Happy coding!👏👏👏

#### Please leave any questions or comments in the PR if you have any doubts or unclarity! You can tag me using @AnselemOdims for a quicker response

**_Note: Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi @rebel216, 

### Good job !!!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Good commit messages
:heavy_check_mark: Descriptive Readme file

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

### N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can get to me on slack  using this handle @mihndim.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**